src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -313,11 +320,39 @@ def _validate_shape(self, extents):
                         "DataSymbols that are part of another symbol shape can"
                         " only be scalar integers, but found '{0}'."
                         "".format(str(dimension)))
```

Comment:
Well we could push through the removal of datasymbols and just check the case where there is a single reference to a datasymbol and in this case check that this datasymbol is an integer. We would then be no worse off than before.

Reply:
Yes it can be done here if you are up for it.

> check the case where there is a single reference

More than that, any reference in the DataNode will be checked if you put the conditions inside the `for reference in references` below.

Reply:
OK, I've done that. The change in the logic was small but I had to change lots of examples. The reason for this is that there is an existing test for references which ensures that if any associated symbols are local, they must be a parameter, otherwise the reference won't have a value and can't be used for dimensioning an array. Moving from a Symbol to a Reference containing the Symbol caused failure in all the test examples where we dimension arrays from local variables (there were lots in the nemo tests).

Note, I decided to only check the datatype for a single reference, rather than looking at all references as you suggest. The reason for this is that we could in theory have a value with a different type that we convert to an integer using NINT, or some similar intrinsic. Therefore I thought I'd keep it simple for the moment.